### ISSUE TITLE
2D convolution with homemade kernel 

### ISSUE THREAD

**Issue by axelcournac** [NVIDIA/MinkowskiEngine#179]
Hi all! Thanks a lot for sharing this library. We want to use it to do sparse 2D convolution with homemade kernel. I put here the code, we don't have the expected results as with the initial conv2D of Pytorch. I would be very grateful if you could tell us what parameters are missing or are to be changed. Best
```python
import numpy
import torch
import torch.nn as nn
import MinkowskiEngine as ME

# input data and kernel:
arr = numpy.zeros((8, 8), dtype=float)
arr[:,:4]  = 4.0
arr[:,4:]  = 1.0

ker = numpy.zeros((3, 3), dtype=float)
ker[:,0] = -1.0
ker[:,2] = 1.0

# Conversion into tensor objects: 
tensor_arr = torch.from_numpy(arr)
tensor_ker = torch.from_numpy(ker)
tensor_arr = torch.unsqueeze(tensor_arr, 0)        ## Yes, you need two invocations
tensor_arr = torch.unsqueeze(tensor_arr, 0)        ##       of unsqueeze()
tensor_ker = torch.unsqueeze(tensor_ker, 0)        ## Yes, you need two invocations
tensor_ker = torch.unsqueeze(tensor_ker, 0)        ##       of unsqueeze()

# 2D convolution with Pytorch class
conv = nn.Conv2d(in_channels=1, 
                 out_channels=1, 
                 kernel_size=3, 
                 bias=False) 
conv.weight = torch.nn.Parameter(tensor_ker)
output = conv(tensor_arr)

"""
output
Out[62]: 
tensor([[[[ 0.,  0., -9., -9.,  0.,  0.],
          [ 0.,  0., -9., -9.,  0.,  0.],
          [ 0.,  0., -9., -9.,  0.,  0.],
          [ 0.,  0., -9., -9.,  0.,  0.],
          [ 0.,  0., -9., -9.,  0.,  0.],
          [ 0.,  0., -9., -9.,  0.,  0.]]]], dtype=torch.float64,
       grad_fn=<ThnnConv2DBackward>)
"""

# 2D convolution with Minkowsky
# reformating the input data:
def to_sparse_coo(data):
    # An intuitive way to extract coordinates and features
    coords, feats = [], []
    for i, row in enumerate(data):
        for j, val in enumerate(row):
            if val != 0:
                coords.append([i, j])
                feats.append([val])
    return torch.LongTensor(coords), torch.FloatTensor(feats)

coords0, feats0 = to_sparse_coo(arr)
coords, feats = ME.utils.sparse_collate(coords=[coords0], feats=[feats0])
ME_tensor_arr = ME.SparseTensor(coords=coords, feats=feats)

conv = ME.MinkowskiConvolution(
    in_channels=1,
    out_channels=1,
    kernel_size= 3, 
    stride=1,
    dilation=1,
    has_bias=False,
    dimension=2)
conv.kernel = torch.nn.Parameter(tensor_ker) 
ME_tensor_arr.double()

output = conv(ME_tensor_arr)
output.dense()

'''
Out[19]: 
(tensor([[[[8.1890e+190, 8.1890e+190, 8.1890e+190, 8.1890e+190, 2.0472e+190,
            2.0472e+190, 2.0472e+190, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 2.0472e+190],
           [1.1559e+205, 1.1559e+205, 1.1559e+205, 2.8898e+204, 2.8898e+204,
            2.8898e+204, 2.8898e+204, 3.2329e+160]],
 
          [[7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [7.2703e+228, 7.2703e+228, 7.2703e+228, 4.5439e+228, 1.8176e+228,
            1.8176e+228, 1.8176e+228, 9.0879e+227],
           [3.6352e+228, 3.6352e+228, 3.6352e+228, 3.6352e+228, 9.0879e+227,
            9.0879e+227, 9.0879e+227, 9.0879e+227]],
 
          [[2.6266e+200, 2.6266e+200, 2.6266e+200, 6.5664e+199, 6.5664e+199,
            6.5664e+199, 6.5664e+199, 2.2066e+161],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199],
           [5.2531e+200, 5.2531e+200, 5.2531e+200, 3.2832e+200, 1.3133e+200,
            1.3133e+200, 1.3133e+200, 6.5664e+199]]]], dtype=torch.float64,
        grad_fn=<CopySlices>),
 tensor([[0, 0]], dtype=torch.int32),
 tensor([1, 1], dtype=torch.int32))
'''
```

**Comment by chrischoy** [comment]
```python
conv = ME.MinkowskiConvolution(
    in_channels=1,
    out_channels=1,
    kernel_size= 3, 
    stride=1,
    dilation=1,
    has_bias=False,
    dimension=2)
conv.kernel[:, 0, 0] = tensor_ker.T[:].flatten() # important!
ME_tensor_arr.double()

print(conv(ME_tensor_arr).dense())

"""
(tensor([[[[ 8.,  0.,  0., -6., -6.,  0.,  0., -2.],
           [12.,  0.,  0., -9., -9.,  0.,  0., -3.],
           [12.,  0.,  0., -9., -9.,  0.,  0., -3.],
           [12.,  0.,  0., -9., -9.,  0.,  0., -3.],
           [12.,  0.,  0., -9., -9.,  0.,  0., -3.],
           [12.,  0.,  0., -9., -9.,  0.,  0., -3.],
           [12.,  0.,  0., -9., -9.,  0.,  0., -3.],
           [ 8.,  0.,  0., -6., -6.,  0.,  0., -2.]]]], dtype=torch.float64,
        grad_fn=<CopySlices>),
 tensor([[0, 0]], dtype=torch.int32),
 tensor([1, 1], dtype=torch.int32))
"""
```

The values are different for boundaries. To get the same value, use the padding for pytorch 2D conv.
